core/src/main/scala/org/http4s/QueryParam.scala

```diff
@@ -91,33 +131,56 @@ object QueryParamDecoder {
       ).toValidatedNel
   }
 
-  def decodeBy[T, U: QueryParamDecoder](f: U => T): QueryParamDecoder[T] = new QueryParamDecoder[T] {
-    def decode(value: QueryParameterValue): ValidatedNel[ParseFailure, T] =
-      QueryParamDecoder[U].decode(value) map f
-  }
-
-
-  implicit val booleanQueryParamDecoder: QueryParamDecoder[Boolean] =
+  /** QueryParamDecoder is a covariant functor. */
+  implicit val FunctorQueryParamDecoder: Functor[QueryParamDecoder] =
+    new Functor[QueryParamDecoder] {
+      override def map[A, B](fa: QueryParamDecoder[A])(f: A => B) =
+        fa.map(f)
+    }
+
+  /** QueryParamDecoder is a MonoidK. */
+  implicit val MonoidKQueryParamDecoder: MonoidK[QueryParamDecoder] =
+    new MonoidK[QueryParamDecoder] {
+      def empty[A] =
+        nothingQueryParamDecoder.widen[A]
+      def combineK[A](a: QueryParamDecoder[A], b: QueryParamDecoder[A]) =
+        a.orElse(b)
+    }
+
+  @deprecated("Use QueryParamEncoder[T].map(f)", "0.17")
+  def decodeBy[U, T](f: T => U)(
+    implicit qpd: QueryParamDecoder[T]
+  ): QueryParamDecoder[U] =
+    qpd.map(f)
+
+  /** A decoder that always fails. */
+  implicit lazy val nothingQueryParamDecoder: QueryParamDecoder[Nothing] =
+    new QueryParamDecoder[Nothing] {
+      override def decode(value: QueryParameterValue) =
+        Validated.invalidNel(ParseFailure("Failed.", "Nothing decoder always fails."))
+    }
+
+  implicit lazy val booleanQueryParamDecoder: QueryParamDecoder[Boolean] =
```

Comment:
What are we getting out of laziness on these?  laziness and some CI lockups have given me a twitch.